src/Microsoft.Extensions.Primitives/CompositeChangeToken.cs

```diff
@@ -0,0 +1,143 @@
+            {
+                if (ChangeTokens[i].ActiveChangeCallbacks)
+                {
+                    ActiveChangeCallbacks = true;
+                    break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the list of <see cref="IChangeToken"/> which compose the current <see cref="CompositeChangeToken"/>.
+        /// </summary>
+        public IReadOnlyList<IChangeToken> ChangeTokens { get; }
+
+        /// <inheritdoc />
+        public IDisposable RegisterChangeCallback(Action<object> callback, object state)
+        {
+            EnsureCallbacksInitialized();
+            if (!_cancellationTokenSource.IsCancellationRequested)
+            {
+                return _cancellationTokenSource.Token.Register(callback, state);
+            }
+            else
+            {
+                return NullDisposable.Singleton;
+            }
+        }
+
+        /// <inheritdoc />
+        public bool HasChanged
+        {
+            get
+            {
+                if (_cancellationTokenSource.Token.IsCancellationRequested)
+                {
+                    return true;
+                }
+
+                for (var i = 0; i < ChangeTokens.Count; i++)
+                {
+                    if (ChangeTokens[i].HasChanged)
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+        }
+
+        /// <inheritdoc />
+        public bool ActiveChangeCallbacks { get; }
+
+        private void EnsureCallbacksInitialized()
+        {
+            if (_registeredCallbackProxy)
+            {
+                return;
+            }
+
+            lock (_callbackLock)
+            {
+                if (_registeredCallbackProxy)
+                {
+                    return;
+                }
+
+                _cancellationTokenSource = new CancellationTokenSource();
+                _disposables = new List<IDisposable>();
+                for (var i = 0; i < ChangeTokens.Count; i++)
+                {
+                    if (ChangeTokens[i].ActiveChangeCallbacks)
+                    {
+                        var disposable = ChangeTokens[i].RegisterChangeCallback(_onChangeDelegate, this);
+                        _disposables.Add(disposable);
+                    }
+                }
+                _registeredCallbackProxy = true;
+            }
+        }
+
+        private static void OnChange(object state)
+        {
+            var compositeChangeTokenState = (CompositeChangeToken)state;
+            lock (compositeChangeTokenState._callbackLock)
+            {
+                try
+                {
+                    compositeChangeTokenState._cancellationTokenSource.Cancel();
+                }
+                catch
+                {
+                }
+            }
+
+            if (compositeChangeTokenState._disposables != null)
```

Comment:
Don't think you need the check here. The callback is only ever registered in the same code path that `_disposables` is initialized. Maybe just do a `Debug.Assert(compositeChangeTokenState._disposables != null);`